add support to run tests with different user permissions

christianvogt opened the pull request:
<!--- If this is a non-code change, this template is not required; reference any issues or top-level descriptions as needed -->
<!--- All code change PRs should relate to an issue, reference it here; see example below -->
<!--- Closes: #123 -->
https://issues.redhat.com/browse/RHOAIENG-5259

## Description
<!--- Describe your changes in detail; the what, the why, any findings, etc -->
<!--- Include any screenshots of changed UI; Include any gifs if it was a flow / UX change -->
Added utility to run tests as users with various permissions:
- disallowed user: no access to load dashboard
- cluster admin: full access to cluster
- project admin: full access specific projects
- edit user: full access to specific projects without ability to create role bindings

Optionally a user may also be a self provisioner (true by default). Non self provisioners cannot create new projects.

Default user for all tests is now a project admin.

Added some small tests that use the various users as well as updated breaking tests.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Run cypress tests.

## Test Impact
<!--- What tests have you done to covert implemented functionality -->
<!--- If tests are not applicable, explain why here -->
N/A

## Request review criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

Self checklist (all need to be checked):
- [x] The developer has manually tested the changes and verified that the changes work
- [x] Commits have been squashed into descriptive, self-contained units of work (e.g. 'WIP' and 'Implements feedback' style messages have been removed)
- [x] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [x] The developer has added tests or explained why testing cannot be added (unit or cypress tests for related changes)

If you have UI changes: 
<!--- You can ignore these if you are doing manifest, backend, internal logic, etc changes; aka non-UI / visual changes -->
- [ ] Included any necessary screenshots or gifs if it was a UI change.
- [ ] Included tags to the UX team if it was a UI/UX change (find relevant UX in the [SMEs](https://github.com/opendatahub-io/odh-dashboard/tree/main/docs/smes.md) section).

After the PR is posted & before it merges:
- [ ] The developer has tested their solution on a cluster by using the image produced by the PR to `main`

andrewballantyne commented on frontend/src/__tests__/cypress/cypress/e2e/Application.cy.ts:
```diff
@@ -0,0 +1,38 @@
+  });
+
+  it('should show Settings nav section for cluster admins', () => {
+    asClusterAdminUser();
```
Can I suggest we either have a test for "cluster admin" or "product admin" -- Cluster Admin sounds like `cluster-admin` and that's technically different than just the admins of our product.

christianvogt replied:
ah yes. 
Having a distinct difference between product admin vs cluster admin is key. In fact almost all usage of cluster admin I did should be product admin.

andrewballantyne commented on frontend/src/__tests__/cypress/cypress/utils/users.ts:
```diff
@@ -0,0 +1,75 @@
+/**
+ * Cluster admins have access to perform any action on resource in every project.
+ */
+export const asClusterAdminUser = (): void => {
```
Ah I see -- you're aiming explicitly at `cluster-admin` -- recommend adding `asProductAdmin` as well. Which should not grant access to any project but will the settings stuff. Not sure if that's possible with this current infra atm.

christianvogt replied:
I will add `asProductAdmin` as suggested.

christianvogt commented:
Rebased.
Added user `asProductAdminUser`.
Moved core intercept initialization for k8s objects from `e2e.ts` to `users.ts`. Expectation now is that tests must establish a user before applying their own test specific intercepts as default intercepts are user specific. `e2e.ts` will still establish the default user as a project admin.
